## **Capstone Session 1: Data Import and Cleaning Tasks**

### **GOAL**
Develop a comprehensive solution for data aggregation, wrangling, and visualization using a healthcare dataset for the Aura platform. The goal is to effectively manage and process complex healthcare data to enable insightful analysis and enhance data-driven decision-making capabilities within Aura.

##### **1. Import relevant Python libraries necessary for Python programming and Numpy**

In [6]:
import pandas as pd
import numpy as np


#### **2. Import the CSV file – NSMES1988.csv into a dataframe.**

In [7]:
df = pd.read_csv('NSMES1988.csv')

####  **3. Inspect the data and report the details from physical inspection – rows, columns, data types etc.**

In [8]:
# Grab the first 5 rows of the DataFrame
df.head()

,Unnamed: 0,visits,nvisits,ovisits,novisits,emergency,hospital,health,chronic,adl,region,age,gender,married,school,income,employed,insurance,medicaid
0,1,5,0,0,0,0,1,average,2,normal,other,6.9,male,yes,6,2.8810,yes,yes,no
1,2,1,0,2,0,2,0,average,2,normal,other,7.4,female,yes,10,2.7478,no,yes,no
2,3,13,0,0,0,3,3,poor,4,limited,other,6.6,female,no,10,0.6532,no,no,yes
3,4,16,0,5,0,1,1,poor,2,limited,other,7.6,male,yes,3,0.6588,no,yes,no
4,5,3,0,0,0,0,0,average,2,limited,other,7.9,female,yes,6,0.6588,no,yes,no


In [9]:
# Give us info on rows, columns and data types
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4406 entries, 0 to 4405
Data columns (total 19 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  4406 non-null   int64  
 1   visits      4406 non-null   int64  
 2   nvisits     4406 non-null   int64  
 3   ovisits     4406 non-null   int64  
 4   novisits    4406 non-null   int64  
 5   emergency   4406 non-null   int64  
 6   hospital    4406 non-null   int64  
 7   health      4406 non-null   object 
 8   chronic     4406 non-null   int64  
 9   adl         4406 non-null   object 
 10  region      4406 non-null   object 
 11  age         4406 non-null   float64
 12  gender      4406 non-null   object 
 13  married     4406 non-null   object 
 14  school      4406 non-null   int64  
 15  income      4406 non-null   float64
 16  employed    4406 non-null   object 
 17  insurance   4406 non-null   object 
 18  medicaid    4406 non-null   object 
dtypes: float64(2), int64(9), ob

#### **4. Find out if the data is clean or if the data has missing values.**

In [11]:
# Find out how many missing values are in each column
df.isna().sum()

Unnamed: 0    0
visits        0
nvisits       0
ovisits       0
novisits      0
emergency     0
hospital      0
health        0
chronic       0
adl           0
region        0
age           0
gender        0
married       0
school        0
income        0
employed      0
insurance     0
medicaid      0
dtype: int64

#### **5. Comment on the data types, their values and range, specifically on age and income columns.**

In [ ]:
# Show the DataFrame's column names before summarizing the numeric fields age and income
print(df.columns)
df[["age","income"]].describe()

Index(['Unnamed: 0', 'visits', 'nvisits', 'ovisits', 'novisits', 'emergency',
       'hospital', 'health', 'chronic', 'adl', 'region', 'age', 'gender',
       'married', 'school', 'income', 'employed', 'insurance', 'medicaid'],
      dtype='object')


,age,income
count,4406.000000,4406.000000
mean,7.402406,2.527132
std,0.633405,2.924648
min,6.600000,-1.012500
25%,6.900000,0.912150
50%,7.300000,1.698150
75%,7.800000,3.172850
max,10.900000,54.835100


#### **6. Export the data to JSON as NSMES1988.json format file and view and enter your comments**

In [ ]:
df.to_json("NSMES1988.json",orient="records",indent=4)

#### **7. Perform memory information on the data and recommend what non-default data types would you recommend to optimize memory settings for the dataframe.**

In [40]:
print(df.nunique())

#Print out the different data types in the DataFrame
print(df.info(memory_usage="deep=True"))

#create a copy of the DataFrame to work with and call it df_downsized
df_downsized = df.copy()

# Downcast the integer columns to the smallest possible integer type to save memory
integer_columns = df_downsized.select_dtypes(include=["int"]).columns
df_downsized[integer_columns] = df_downsized[integer_columns].apply(pd.to_numeric, downcast="integer")

#downcast the object columns to the smallest possible object type to save memory
object_columns = df_downsized.select_dtypes(include=["object"]).columns
df_downsized[object_columns] = df_downsized[object_columns].apply(lambda x: x.astype("category") if x.nunique() / len(x) < 0.5 else x)

print(df_downsized.info(memory_usage="deep=True"))


# print out the memory usage of each dataframe but do not break it down by column
print(f"\n\nOriginal DataFrame memory usage: {df.memory_usage(index=False, deep=True).sum()/1024**2:.2f} MB")
print(f"Downsized DataFrame memory usage: {df_downsized.memory_usage(index=False, deep=True).sum()/1024**2:.2f} MB")

Unnamed: 0    4406
visits          60
nvisits         51
ovisits         37
novisits        35
emergency       11
hospital         9
health           3
chronic          9
adl              2
region           4
age             36
gender           2
married          2
school          19
income        3015
employed         2
insurance        2
medicaid         2
dtype: int64
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4406 entries, 0 to 4405
Data columns (total 19 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  4406 non-null   int64  
 1   visits      4406 non-null   Int64  
 2   nvisits     4406 non-null   int64  
 3   ovisits     4406 non-null   int64  
 4   novisits    4406 non-null   int64  
 5   emergency   4406 non-null   int64  
 6   hospital    4406 non-null   int64  
 7   health      4406 non-null   object 
 8   chronic     4406 non-null   int64  
 9   adl         4406 non-null   object 
 10  region      4406 non-n

##### **MEMORY ANALYSIS**
After analyzing the dataframe, changing the default datatypes appears to be the biggest way to reduce memory storage. Columns that are **int64** should be reduced to int8 or int16 data type because the data contained in those columns are not large numbers requiring int64. 

Similarly, the columns that are of **object** data type should be changed to category data type. 

These 2 changes reduce the dataframe size significantly. The original size was **2.43MB**, but with the suggested data type changes, the size drops to **0.16MB** 

Those columns that are float data type should not be touched and remain float so as to keep their precision

#### **8. Recommended changes to the dataframe before attempting detailed data analysis**

As recommended in the previous section, the following data type chages should be performed

**Unnamed**: int64 --> int16<br>
**visits**: int64 --> int<br>
**nvisits**&nbsp;&nbsp;&nbsp;int64<br>
**ovisits**     int64<br>
**novisits**    int64<br>
**emergency**   int64<br>
**hospital**    int64<br>
**health**      object<br>
**chronic**     int64<br>
**adl**         object<br>
**region**      object<br>
**age**         float64<br>
**gender**      object<br> 
**married**     object<br> 
**school**      int64<br>
**income**      float64<br>
**employed**    object<br>
**insurance**   object<br> 
**medicaid**    object<br> 

dtypes: Int64(1), float64(2), int64(8), object(8)
memory usage: 658.4+ KB

RangeIndex: 4406 entries, 0 to 4405
Data columns (total 19 columns):
 #   Column      Non-Null Count  Dtype   
---  ------      --------------  -----   
 0   Unnamed: 0  4406 non-null   int16   
 1   visits      4406 non-null   Int8    
 2   nvisits     4406 non-null   int8    
 3   ovisits     4406 non-null   int16   
 4   novisits    4406 non-null   int16   
 5   emergency   4406 non-null   int8    
 6   hospital    4406 non-null   int8    
 7   health      4406 non-null   category
 8   chronic     4406 non-null   int8    
 9   adl         4406 non-null   category
 10  region      4406 non-null   category
 11  age         4406 non-null   float64 
 12  gender      4406 non-null   category
 13  married     4406 non-null   category
 14  school      4406 non-null   int8    
 15  income      4406 non-null   float64 
 16  employed    4406 non-null   category
 17  insurance   4406 non-null   category
 18  medicaid    4406 non-null   category


#### **9. Export the data frame as a new CSV file NSMES1988new.csv and store it in the local space for likely use in other assignments.**

In [39]:
#export the downsized DataFrame to a new CSV file called NSMES1988new.csv
df_downsized.to_csv("NSMES1988new.csv", index=False)

In [ ]:
df.info(memory_usage="deep=True")
df_downsized.info(memory_usage="deep=True")



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4406 entries, 0 to 4405
Data columns (total 19 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  4406 non-null   int64  
 1   visits      4406 non-null   Int64  
 2   nvisits     4406 non-null   int64  
 3   ovisits     4406 non-null   int64  
 4   novisits    4406 non-null   int64  
 5   emergency   4406 non-null   int64  
 6   hospital    4406 non-null   int64  
 7   health      4406 non-null   object 
 8   chronic     4406 non-null   int64  
 9   adl         4406 non-null   object 
 10  region      4406 non-null   object 
 11  age         4406 non-null   float64
 12  gender      4406 non-null   object 
 13  married     4406 non-null   object 
 14  school      4406 non-null   int64  
 15  income      4406 non-null   float64
 16  employed    4406 non-null   object 
 17  insurance   4406 non-null   object 
 18  medicaid    4406 non-null   object 
dtypes: Int64(1), float64(2), in